Connect "Contract interaction" transaction

ArmandoGraterol opened the pull request:
**To test:**

- Run `npm run dev:heavy -- --skip-webpack`
- Follow the instructions in https://github.com/JoinColony/colonyDapp/pull/3765 to get the foreign bridge running. Make sure you have the latest changes from `maint/cross-chain-safe`.
- Add the env variable values to your .env and then start webpack (`npm start webpack` in `/colonyDapp`).
- Add a safe. For example: `0x4cb09344de5bCCD45F045c5Defa0E0452869FF0f` (Rinkeby)
- Go to safe control, select "contract interaction", and fill out the form. Use any verified contract that is part of the chain of the selected safe.
- Confirm the transaction.
- In the terminal window running the foreign bridge, you should see a confirmation that funds have been transferred out of the safe to a 0 address. In the console in the browser, you should see proof that the bridging transaction was successful.

NOTE: 
- In the local env, it's not expected that the safe you added is used to transfer the funds, or that the function you select in the form is the one used for the transaction. 
- A local safe with 100 test tokens in it is going to be used.
- I've used `transferFrom` out of convenience, the process of testing this transaction is quite similar to testing the transfer funds transaction with the exception that the transfer funds version uses `transfer`  


Resolves #3726

jakubcolony reviewed (approved):
LGTM! 👍

After a few issues with setting up I managed to get it running:
<img width="494" alt="image" src="https://user-images.githubusercontent.com/112586815/194172255-9fa0b427-5595-49af-befd-ae8df08bc8aa.png">
<img width="568" alt="image" src="https://user-images.githubusercontent.com/112586815/194172287-5ff64a00-4c00-437c-a4be-39466ae7547d.png">

And was redirected to:
<img width="940" alt="image" src="https://user-images.githubusercontent.com/112586815/194172330-71cfb823-d84c-44ee-99bc-ecdd943cdb90.png">

willm30 commented on src/modules/dashboard/sagas/utils/safeHelpers.ts:
```diff
@@ -280,6 +281,92 @@ export const getTransferFundsData = async (
+};
+
+const isArrayParameter = (parameter: string): boolean =>
+  /(\[\d*])+$/.test(parameter);
```
How come you're only checking for arrays containing numeric digits? If I understand correctly, this is intended to test whether the start of the string is '[' and the end ']' (i.e. an array)? In which case, the following might be simpler: 
`parameter[0] + parameter[parameter.length - 1] === '[]'`

ArmandoGraterol replied:
You are right, I don't remember why did I include the `\d*`

willm30 commented on src/modules/dashboard/sagas/utils/safeHelpers.ts:
```diff
@@ -280,6 +281,92 @@ export const getTransferFundsData = async (
+): (string | number)[] | null => {
+  try {
+    const arrayResult = JSON.parse(parameter);
+    return arrayResult.map((value) => {
```
I believe JSON.parse also introduces rounding errors for large numbers. E.g. 2 ** 255 - 1 (the largest safe int) will be turned into 
`5.78960446186581e+76`, which does not `===` 2 ** 255 - 1. 

Whereas `getArrayFromString` will give you an array of the numbers as strings, so you don't need to convert to a big number and then back to a string.

willm30 commented on src/modules/dashboard/sagas/utils/safeHelpers.ts:
```diff
@@ -280,6 +281,92 @@ export const getTransferFundsData = async (
+  signature: string,
+  transaction: Record<string, any>,
+) => ({ name = '' }) => {
+  if (isArrayParameter(transaction[name || signature])) {
```
Do we store the signature as a key in the SafeTransaction object? I've not come across this, I've only noticed that we store the function name as the key and the input data as the value.

ArmandoGraterol replied:
True, I think this is a leftover from when I was testing some things

ArmandoGraterol replied:
Gonna have to update this anyways with the merge of #3952

willm30 commented on src/modules/dashboard/sagas/utils/safeHelpers.ts:
```diff
@@ -280,6 +281,92 @@ export const getTransferFundsData = async (
+    ];
+
+    const transactionValues = onLocalDevEnvironment
+      ? { ...transaction, src: safeAddress, dst: AddressZero, wad: 1 }
```
Perhaps worth a comment here that `src`, `dst` and `wad` are the inputs to the `transferFrom` function

willm30 reviewed (commented):
Working for me! 

Left a question about the way we're turning a string array into a javascript array.

Julianahlk reviewed (commented):
When having interactions with functions that do not contain array data, the transactions go through smoothly:
![simpleinteraction](https://user-images.githubusercontent.com/71563622/194403320-4e093d72-6922-42db-84b2-17a5a09eb7aa.gif)
<img width="943" alt="Screen Shot 2022-10-06 at 9 39 11 PM" src="https://user-images.githubusercontent.com/71563622/194403820-291bace0-518e-4de9-aac1-9d59e6fbc5e4.png">

But I am having some issues when interacting with the following array data:  
![arraydata](https://user-images.githubusercontent.com/71563622/194405618-4c9eba88-e3f2-4cb8-94c9-183fe6e9e787.gif)
![arrays2](https://user-images.githubusercontent.com/71563622/194405628-84778bba-80be-4772-9f75-2f2501086c6a.gif)

If there is just one single array[] there is not issue:
![onearray](https://user-images.githubusercontent.com/71563622/194406065-63490e51-6b29-48d1-bc14-63b9986ac041.gif)

ArmandoGraterol commented on src/modules/dashboard/sagas/utils/safeHelpers.ts:
```diff
@@ -280,6 +281,92 @@ export const getTransferFundsData = async (
+  parameter: string,
+): (string | number)[] | null => {
+  try {
+    const arrayResult = JSON.parse(parameter);
```
Sure, I'll update and use that one then

chinins reviewed (approved):
Looking good!

ArmandoGraterol commented:
@Julianahlk Hey! Can you test again? I think this is some bug in the form itself rather than the code I added to connect the transaction, at least it looks like it from the gif you added, but I've rebased and fixed some weird stuff in the validation. Also, can you provide me the contract address you used to get the param of the array inside an array?

Julianahlk commented:
> @Julianahlk Hey! Can you test again? I think this is some bug in the form itself rather than the code I added to connect the transaction, at least it looks like it from the gif you added, but I've rebased and fixed some weird stuff in the validation. Also, can you provide me the contract address you used to get the param of the array inside an array?

I just tested again and I am getting the same `Error: invalid BigNumber`. It does look like this issue is more related to the validation and not your code. Here is the address I am using for the multidimensional array: 
`0x44762835EeB462fCe0150935593Fd119A935B16a`
Function name: unstake

And this one has single arrays:
`0xE65841eb30548E7be3FFB9e496DB1B540C5eD8C6`
Function name: addLeaderboardEntry

willm30 commented:
@Julianahlk I've opened an issue for nested array validation: https://github.com/JoinColony/colonyDapp/issues/3986 . Good catch.

willm30 reviewed (commented):
@ArmandoGraterol Receiving the following error. The  `const = paramName` in `extractMethodArgs` doesn't account for the param names used in the token contract in testing:

![arry](https://user-images.githubusercontent.com/64402732/195548555-4d3ba3ce-95f7-44a8-9bfb-d803f8d5e1b9.png)
![extractMethodArgs](https://user-images.githubusercontent.com/64402732/195549319-05edde5b-0801-4c6e-b197-1cfa27d4ce48.png)

ArmandoGraterol commented:
> @ArmandoGraterol Receiving the following error. The `const = paramName` in `extractMethodArgs` doesn't account for the param names used in the token contract in testing:
> 
> ![arry](https://user-images.githubusercontent.com/64402732/195548555-4d3ba3ce-95f7-44a8-9bfb-d803f8d5e1b9.png) ![extractMethodArgs](https://user-images.githubusercontent.com/64402732/195549319-05edde5b-0801-4c6e-b197-1cfa27d4ce48.png)

Damn, sorry about that. I was tunnel visioned into the other bugs :face_with_head_bandage:

willm30 reviewed (approved):
Nice one, working for me👍
